System/Random.hs

```diff
@@ -330,18 +335,68 @@ genRandom = randomM PureGen
 splitGen :: (MonadState g m, RandomGen g) => m g
 splitGen = state split
 
-runStateGen :: RandomGen g => g -> State g a -> (a, g)
-runStateGen = flip runState
+runGenState :: RandomGen g => g -> State g a -> (a, g)
+runGenState = flip runState
 
-runStateGen_ :: RandomGen g => g -> State g a -> a
-runStateGen_ g = fst . flip runState g
+runGenState_ :: RandomGen g => g -> State g a -> a
+runGenState_ g = fst . flip runState g
 
-runStateTGen :: RandomGen g => g -> StateT g m a -> m (a, g)
-runStateTGen = flip runStateT
+runGenStateT :: RandomGen g => g -> StateT g m a -> m (a, g)
+runGenStateT = flip runStateT
 
-runStateTGen_ :: (RandomGen g, Functor f) => g -> StateT g f a -> f a
-runStateTGen_ g = fmap fst . flip runStateT g
+runGenStateT_ :: (RandomGen g, Functor f) => g -> StateT g f a -> f a
+runGenStateT_ g = fmap fst . flip runStateT g
 
+-- | This is a wrapper around a pure generator that can be used in an effectful environment.
+-- It is safe in presence of concurrency since all operations are performed atomically.
+--
+-- @since 1.2
+newtype PrimGen s g = PrimGen (MutVar s g)
+
+instance (s ~ PrimState m, PrimMonad m, RandomGen g) =>
+         MonadRandom (PrimGen s g) m where
+  type Seed (PrimGen s g) = g
+  restore = fmap PrimGen . newMutVar
+  save (PrimGen gVar) = readMutVar gVar
+  uniformWord32R r = atomicPrimGen (genWord32R r)
+  uniformWord64R r = atomicPrimGen (genWord64R r)
+  uniformWord8 = atomicPrimGen genWord8
+  uniformWord16 = atomicPrimGen genWord16
+  uniformWord32 = atomicPrimGen genWord32
+  uniformWord64 = atomicPrimGen genWord64
+  uniformByteArray n = atomicPrimGen (genByteArray n)
+
+-- | Apply a pure operation to generator atomically.
+atomicPrimGen :: PrimMonad m => (g -> (a, g)) -> PrimGen (PrimState m) g -> m a
+atomicPrimGen op (PrimGen gVar) =
+  atomicModifyMutVar' gVar $ \g ->
+    case op g of
+      (a, g') -> (g', a)
+
+
+-- | Split `PrimGen` into atomically updated current generator and a newly created that is
+-- returned.
+--
+-- @since 1.2
+splitPrimGen ::
+     (RandomGen g, PrimMonad m)
+  => PrimGen (PrimState m) g
+  -> m (PrimGen (PrimState m) g)
+splitPrimGen = atomicPrimGen split >=> restore
+
+runPrimGenST :: RandomGen g => g -> (forall s . PrimGen s g -> ST s a) -> (a, g)
+runPrimGenST g action = runST $ do
+  primGen :: PrimGen s g <- restore g
+  res <- action primGen
+  g' <- save primGen
+  pure (res, g')
+
+runPrimGenIO :: (RandomGen g, MonadIO m) => g -> (PrimGen RealWorld g -> m a) -> m (a, g)
+runPrimGenIO g action = do
+  primGen :: PrimGen s g <- liftIO $ restore g
```

Comment:
```suggestion
  primGen <- liftIO $ restore g
```

Reply:
Sorry, this has already been fixed in: https://github.com/idontgetoutmuch/random/pull/17/files#diff-4a522d098452ebfc9be2d6291588f5bcL396

Reply:
In case you wanna experiment with MutGen and `Prim` instance for SMGen here is a branch that has it: https://github.com/lehins/splitmix/tree/new-random

Also the benchmarks that I ran are here: https://github.com/lehins/splitmix/tree/new-random